ucb-rit/coldfront
# Generalize allowance-related handling for LRC; replace hard-coded strings; hide deployment-specific functionality

Refs #383
Fixes #384
Fixes #414

**Changes**
- Added the concept of a computing allowance, modeled as a `Resource`. This replaces hard-coded `allocation_type` strings (e.g., `SavioProjectAllocationRequest.FCA`, etc.) with values stored in the database.
    - Each `Resource` has multiple `ResourceAttributeTypes` (see `add_allowance_defaults` for examples):
        - `code` (the three letter prefix before a project name)
        - `name_short` (a short name for the allowance)
        - `name_long` (a long name for the allowance)
        - `Service Units` (the default number of service units to assign to an `Allocation` to this allowance, if applicable)
    - Added a class `ComputingAllowanceInterface` for fetching attribute values or the associated `Resource` objects.
    - Added a class `ComputingAllowance` to wrap around computing allowance `Resources` and provide answers to deployment-specific questions.
        - E.g., does this allowance have infinite service units? Is it a periodic allowance? Is it yearly? Can it be pooled? Does it require an MOU?
            - Replaced checks against allocation types/prefixes with these methods.
    - Added a foreign key to the corresponding computing allowance `Resource` to new project requests and `AllocationRenewalRequests`.
    - Replaced `allocation_type` with `computing_allowance` in forms, views, tests, etc.
        - The field was left for now, but could be removed later.
        - Removed `ALLOCATION_TYPE_CHOICES` from `SavioProjectAllocationRequest`.
- Removed hard-coded references to project name prefixes (e.g., `fc_`) in *non-test* code.
- Removed hard-coded references to "Savio".
    - Renamed new project request views to be general.
    - Displayed the correct cluster name in various places.
- Limited the following functionality to BRC:
    - Service Unit Purchase Requests (Project Detail view for Recharge projects, Request Hub, Navbar)
    - New Vector project requests (Create a new project, Request Hub, Navbar)
    - Requirement of an MOU for new Recharge project requests
- Removed all references to the `AllocationAttributeType` named `{CLUSTER_NAME} Allocation Type`, which served no purpose.
- Introduced the following settings:
    - `PRIMARY_CLUSTER_NAME` ("Savio" for BRC, "Lawrencium" for LRC)
        - Replaced hard-coded references to "Savio" with the setting.
        - Added utility methods for deriving the correct "{CLUSTER_NAME} Compute" `Resource` from the setting, so that references to "Savio Compute" could be replaced.
- Removed the following settings:
    - `CO_DEFAULT_ALLOCATION` (now stored in database)
    - `ICA_DEFAULT_ALLOCATION` (now stored in database)
    - `FCA_DEFAULT_ALLOCATION` (now stored in database)
    - `PCA_DEFAULT_ALLOCATION` (now stored in database)
- Introduced the following feature flags:
    - `SERVICE_UNITS_PURCHASABLE` (`True` for BRC, `False` for LRC)
- Introduced the following management commands:
    - `add_allowance_defaults` (create new `Resources` and `ResourceAttributeTypes`)
    - `set_new_project_request_computing_allowances` (set computing allowances for existing new project requests)
    - `set_allocation_renewal_request_computing_allowances` (set computing allowances for existing `AllocationRenewalRequests`)

**How to Test**
- Ensure that the test suite passes on both a BRC-based VM and an LRC-based VM.
    - The Ansible playbook should be re-run first.
- Manual Testing
    - Set `primary_cluster_name` in `main.yml` to "Savio"/"Lawrencium".
    - Re-run the Ansible playbook.
    - Run the following management commands for setting new field values for existing objects:
        ```
        python manage.py set_new_project_request_computing_allowances # --dry_run first
        python manage.py set_allocation_renewal_request_computing_allowances # --dry_run first
        ```
    - Ensure that BRC-specific functionality is hidden from LRC.
    - Test miscellaneous request/approval flows and note any errors.
    - Note any missed references to deployment-specific strings.

**Notes**
- Tests remain highly BRC-specific.
- `SavioProjectAllocationRequest` has not been renamed.
